refactor onoff to separate async notification, standardize terminology, and add functionality

pabigot opened the pull request:
PR tracks changes to #23601 and supplies changes to #23333

This PR separates generic resource management updates in #22853 from the new queued operation API, and adds an extension to the onoff manager intended to satisfy the use case of #22974 without having to change how cancellation is implemented.  It is intended as something like a rebaseable topic branch where the earlier commits are on a path to stabilization but are still impacted by changes in functionality and API due to evolution of later commits.

The PR branch:
* refactors the onoff service to reduce space by putting static configuration data into a separate structure;
* extracts the async notification infrastructure from the on-off service into a self-contained structure that can be used anywhere asynchronous operations may be used;
* recasts the on-off service as a manager that is to be used by services (including API changes to make it more consistent with other managers to come);
* reworks the API to support a change to how cancellation works, providing an ability for the last client waiting for an operation to cancel that operation, using techniques inspired by #23087 and prototyped in #23512;
* reworks and extends API to support handling behavioral issues related to cancelling operations, by providing a monitor facility for synchronous notification of state changes in the underlying service;
* adds an intermediary that converts asynchronous request and release API to synchronous requests and releases with callback notification when the desired state has been reached, which is an alternative approach that avoids cancellation but satisfying the use case of #22974.

nordic-krch committed:
lib: os: onoff: Move transition functions out of service struct

Extracted transition functions from onoff structure to external one
which allows to keep them in flash.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pabigot committed:
sys: add generic asynchronous notification infrastructure

k_poll() for a signal is often desired for notification of completion
of asynchronous operations, but there are APIs where it may be
necessary to invoke "asynchronous" operations from contexts where
sleep is disallowed, or before the kernel has been initialized.
Extract the general notification solution from the on-off service into
a utility that can be used for other APIs.

Also move documentation out to a resource management section.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
lib: os: onoff: generalize and shorten API

The original API was misnamed, as the intent was to provide a manager
that decoupled state management from the service that needed to be
turned on or off.  Update all the names, shortening them where
appropriate removing unncessary internal components like _service.

Also remove some API that misled developers into believing that onoff
managers are normally expected to be exposed directly to consumers.
While this is a use case, in most situations there are service or
client-specific actions that need to be coupled to transition events.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
DNM lib: os: onoff: add API for event-based notification

When functionality that requires an onoff service is affected by
external asynchronous actions, such as plugging and unplugging of
connections, the management of the current state of an asynchronous
service is too complex to handle in individual clients.  Provide an
intermediary that can handle an arbitrary sequence of interleaved
requests and releases, tracking the most recent desired state and
notifying the client only when the state has been reached.

DNM until API and behavior has stabilized to the point where the
depth tracking and PRINTK trace facility can be removed.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

nordic-krch commented on lib/os/onoff_notifier.c:
```diff
@@ -0,0 +1,165 @@
+	}
+
+	if ((rv == 0) && do_change) {
+		rv = initiate_change(np, target_on);
```
shouldn't that be `change_state()`?

nordic-krch replied:
also it's with locked interrupts

nordic-krch reviewed (changes requested):
my main concern is calling onoff manager from locked context.

pabigot commented:
@nordic-krch Yes, the previous implementation had several errors, not unexpected since I hadn't worked on the test suite yet.  The largest was not taking into account the need to release the mutex around two operations, then check for pending state changes after the mutex is complete.  The updated one appears to be correct.

pabigot commented:
Also, as with #22853, some work has to be done to deal with the fact that an onoff service might enter an error state so its reset capability needs to be reachable from the extended API.  This will come soon, for now I want to know whether this satisfies the functional requirements of the USB module's asynchronous request/release events as identified in https://github.com/zephyrproject-rtos/zephyr/issues/22974#issuecomment-591471296.

nordic-krch reviewed (commented):
Few comments added. I haven't yet reviewed `async_notifier`.

nordic-krch commented:
Regarding `async_notifier`: I still believe that such complex solution may not be needed. I want to look into extending onoff manager to support canceling even for last client. In theory it should be possible to move client from list to releaser and return success only to get notification when process is completed with -ECANCELED error (or error reported by the service). That would be cleaner imo.

In that case requesting with unknown client state would look like:
```
err = onoff_cancel(mgr, cli);
if (err ==  -EALREADY) {
  /* cli is in idle, we are in the settled state */
  onoff_init..callback(cli, started_cb);
  err = onoff_request(mgr, cli);
  /* handle potential error */
} else if (err == 0) {
  /* it means that manager was releasing and we managed to cancel it. */
}
```

and releasing looks similar:
```
err = onoff_cancel(mgr, cli);
if (err ==  -EALREADY) {
  /* cli is in idle, we are in the settled state */
  onoff_init_callback(cli, stopped_cb);
  err = onoff_release(mgr, cli);
  /* handle potential error */
} else if (err == 0) {
  /* it means that manager was transitioning to on and we managed to cancel it. */
}
```

In callback we are able to determine state: if `started_cb` results in -ECANCELED it means service is off (for this client). If `stopped_cb` results in -ECANCELED it means service is on (for this client).

The main goal would be to ensure that behavior is the same disregarding other clients requests (now that is not true since last one needs special treatment).

pabigot commented on include/sys/async_notify.h:
```diff
@@ -0,0 +1,352 @@
+	 * async infrastructure.  This field is the sole record of
+	 * success or failure for spin-wait synchronous operations.
+	 */
+	int volatile result;
```
> Also in case of callback notify using fetch result also does not make much sense since by the time it is fetched, async notify may already be used for something else and result is not valid.

Only if the user writes bad code.  Control of the object transfers to the user when the notification completes.  Nothing else is going to touch the content until the same object is given back to the service.

Note that this is not true of the signal result, becajuse the `k_poll_signal` object might be shared between multiple transition result objects, causing a wakeup when the first one completes.  This is another reason why having the result associated with the `async_notify` object rather than hidden away somewhere is useful.

And similarly for callback: it's passed to the callback as a convenience, but the callback may do nothing more than put the result on a queue for later processing.

> does it need to be `volatile`? It seems that when fetching result `flags` is checked and only when it indicates completion then result is read.

The `compiler_barrier()` I added recently in `async_notify_finalize()` shouldn't be necessary; if it's removed then `result` needs to be volatile to guarantee it's set after `flags`.  I'd like to remove that barrier.

tbursztyka replied:
> Note that this is not true of the signal result, because the k_poll_signal object might be shared between multiple transition result objects

That sounds bad code to me to do that, as the one you mentioned earlier. How is the user going to know which objects the signal was about if these are sharing the same k_poll_signal?

nordic-krch replied:
> Note that this is not true of the signal result, because the k_poll_signal object might be shared between multiple transition result objects, causing a wakeup when the first one completes.

If there are a shared signal being used and one task completes but async_notify client did not completed yet. How user can pend on that completion? Only by spining on fetching result?

pabigot replied:
I was going to say "by using `async_notify_fetch_result` to see if the result is ready" but that contradicts my statement in https://github.com/zephyrproject-rtos/zephyr/pull/23229#discussion_r391558073 that the notification isn't complete until a callback is invoked.  So I'm going to have to change my position in that case.

Here, though, it's still valuable.  The other is that, because callbacks must be isr-ok and not sleep, there will be applications where the only thing the callback does is add the object that contains the notification object to a list for later processing in a thread.  In that case something still needs to store the specific result.

nordic-krch replied:
yes, but then i don't think that it will be storing async_notifes rather another structure. Also in that case there can always be a dedicated structure holding async_notify and result. Or even use result field in if spinwait method was used. User can even utilize available space in `flags` to mark given async_notify object as the one holding result.

pabigot replied:
It will probably be adding the object that contains the `async_notify`, as with the onoff manager.

pabigot commented:
> Regarding `async_notifier`: I still believe that such complex solution may not be needed.

Did you mean `onoff_notifier` here?

Note that `onoff_notifier` is far more general and useful than cancellation.  It is expected to handle an arbitrary sequence of request/release items asynchronous to the thing that will use the service once it's turned on.  It works even if neither transition can be cancelled.  It removes all of the management logic from the client, so client code doesn't have to know whether it has an outstanding request that needs to be cancelled.  All the client does is listen for change events.

I accept you've found a case for Nordic clocks where it appears to be safe to synchronously stop an in-progress non-instantaneous start transition, but don't accept that this is likely to be common.  So a solution that works to handle the original use case without involving cancelling in-progress requests and adding client-specific code seems valuable.  In fact it's probably how I'll handle `queued_operation_manager` use of onoff, because the need to support asynchronous reset becomes extremely complex without it.

pabigot commented on lib/os/onoff.c:
```diff
@@ -1,127 +1,81 @@
-	default:
-		__ASSERT_NO_MSG(false);
+	if (cb) {
+		cb(mp, cli, cli->user_data, res);
```
I don't think that argument holds up.

The user specifies the notification method as spin-wait, signal, or callback when the `async_notify` object is initialized.    Notification involving a callback is not complete until the translation function that invokes the callback has invoked that callback.  Prior to invoking the caller hasn't been notified of the completion, so the service retains control of the data object and nobody can have changed the `user_data` pointer.  (I'll update the documentation if that isn't clear.)

The theoretical situation would be a problem only if a client specified a callback notification, then used spin-wait to check for completion and started re-using the structure before the originally requested callback notification was performed.  That's misuse of the API.

pabigot replied:
> I don't think that argument holds up.

https://github.com/zephyrproject-rtos/zephyr/pull/23229#discussion_r391563335 introduces an argument that supports detecting notification before the callback is invoked.  So there'll have to be at least a documentation clarification.  I'm not sure yet if the code will have to be changed.

pabigot replied:
I've updated the documentation to make a distinction between release of the `async_notify` object and release of the containing object, particularly in the case of callback notification.

pabigot committed:
lib: os: onoff: rework to support cancelling in-progress operations

An original design goal of the onoff infrastructure was to ensure that
any errors discovered during transitions would be communicated to at
least one client.  To meet this goal cancellation of the last client
associated with an in-progress transition must be rejected.

This ends up significantly complicating the logic associated with
operation cancellation, and impacted clients that were forced to
maintain state to issue a release or request when the in-progress
transition completed.

Rework transition management so that both request and release
transitions can be accomodated at the same time.  Leave a cancelled
request notification uncompleted, as the client can tell by the
cancellation return value that it has received control of the client
object.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented on include/sys/onoff.h:
```diff
@@ -236,7 +220,7 @@ struct onoff_client;
  * Compatible functions must be isr-callable and non-suspendable.
  *
- * @param srv the service for which the operation was initiated.
+ * @param mp the manager for which the operation was initiated.
```
reverted

nordic-krch commented:
@pabigot to move forward could you take out `onoff notifier` from this PR into another one? That should go in quickly then. That will allow onoff cancel improvement, qop manager and async_seq manager to become merge-able (based on master).

pabigot committed:
lib: os: onoff: rework to make state flags visible to clients

Upcoming changes will inform the clients of the state of the system at
points where it changes.  Make the state-related bits visible to
enable the clients to determine the current state.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
lib: os: onoff: include state at time of completion in notify callback

Upcoming revision of notifications for cancelled operations includes
the possibility that a client originally submitted for a request
operation will instead receive a notification for a completed release
operation.  Although this doesn't solve the problem for spinwait or
signal notifications, add the state-at-completion as a parameter to
the callback notification signature.

Since we're changing things, move the user data parameter to the last
argument, which is its position in most other callback APIs.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
lib: os: onoff: add ability to monitor service state changes

It's possible to have a transition complete with no clients available
to receive the result, which may be a notification of error.
Alternatively a notification that was originally associated with a
release may be converted to a notification associated with a request
due the cancellation of the release.  There is no way to detect this
from spinwait or signal notifications.

Also a module may want to centralize tracking of current state and
error handling for an onoff service that gets used from multiple
contexts.

In support of this provide API to monitor state changes within an
onoff service.  Include notifications when the service transitions out
of stable (ON or OFF) states as well as detection of errors.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

nordic-krch commented:
> @pabigot to move forward could you take out onoff notifier from this PR into another one? 

@pabigot do you plan to do that? It's getting harder to review PR's which contain unrelated changes (async_notify) like #23512

pabigot commented:
> > @pabigot to move forward could you take out onoff notifier from this PR into another one?
> 
> @pabigot do you plan to do that? It's getting harder to review PR's which contain unrelated changes (async_notify) like #23512

No, I don't.  Changes to cancelling and notification impact the whole onoff manager so I want the complete set of changes to that done in one PR.  It's likely #23512 will replace this PR but I haven't decided yet exactly where the cut between draft and real PR will be.  For now, do nothing more with this.

nordic-krch commented:
> Changes to cancelling and notification impact the whole onoff manager 

It does not impact renaming service to manager and extracting async_notify from onoff and moving transistion functions to external struct. So I would like to see those as separate PR which may go in very soon. It seems that onoff functional changes will take longer. 
I'm pushing for that because there is other PR depending on extraction of async notify which is now blocked by onoff functional discussions which are not related to it.

pabigot commented:
Which PR is that, please?

nordic-krch commented:
@pabigot contribution guidelines states: "One general practice we encourage, is to make small, controlled changes." and even title of this PR suggests that there are 3 independent changes.

PR is #23075.

pabigot commented:
> PR is #23075.

I understood that to depend on the queued operation manager in #23333 which is based on this branch.  Until yesterday that was potentially dependent on onoff_notifier, which was thought to be simplifiable by improving cancellation.  Today it isn't, but that could change.  (It does have a syntactic dependence on the notify callback changes motivated by cancellation.)

> @pabigot contribution guidelines states: "One general practice we encourage, is to make small, controlled changes." and even title of this PR suggests that there are 3 independent changes.

They're not independent.  Each one can depend on the earlier ones, depending on what each enhancement provides.  This is a draft because things keep changing and I'm tired of maintaining three branches with pieces that get out of date and are hard to rebase.

If we're done changing the semantics and implementation of cancelling in-progress on-off transitions, then we can start to peel things off.  But I won't be confident of that until we have a solution you and I agree will work for #22974.  From my perspective right now the best solution is onoff_notifier.

nordic-krch commented:
> I understood that to depend on the queued operation manager in #23333 which is based on this branch

no, that is #23371. #23075 is using `async_notify` extracted from onoff.
I don't see how this https://github.com/zephyrproject-rtos/zephyr/pull/23229/commits/c3c70730cdad8f30db0594a77b08627927a61f50 is dependent? it could go to master much faster than the rest. Same with https://github.com/zephyrproject-rtos/zephyr/pull/23229/commits/61b9df39d613c523920d096f5dd57e059f3d7de9 which i had ready in PR but you convinced me to taken it into your PR which won't get merged in near future.

> This is a draft because things keep changing

That applies only to cancel/notifier work and not `refactor onoff to separate async notification, standardize terminology` part of PR.

carlescufi commented on include/sys/async_notify.h:
```diff
@@ -0,0 +1,352 @@
+ * However, services that use this structure may access certain
+ * fields directly.
+ */
+struct async_notify {
```
Yes please, same for other services/managers.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -365,23 +407,24 @@ int onoff_release(struct onoff_service *srv,
-		srv->releaser = cli;
+		set_service_state(mgr, ONOFF_STATE_TO_OFF);
+		set_client_type(cli, ONOFF_CLIENT_RELEASE);
+		sys_slist_append(&mgr->clients, &cli->node);
```
since we have monitors now, then last releaser does not need to be enqueued to notify about stopping result. If that is applied then canceling simplifies a lot since release and request are instantly cancelled. There is probably no need to have client type field for the release. Also that would mean that `onoff_release` could be synchronous and actually anonymous (so also no release cancel, `onoff_cancel` -> `onoff_cancel_request`). Imo, that would be more natural for the client which care when service is on but does not care what happens when service is released. Same applies to monitor, which also seems more natural: one (or more) maintainer of the service which tries to help (applies reset) when something goes wrong. In that approach, `onoff_reset` could also be anonymous as monitor gets notified when service is restored.

pabigot replied:
That's a pretty significant architectural change which suggest additional changes. But it has some potential; I'll think about this and look into it sometime next week.

pabigot commented:
onoff solution has been completedly redesigned, this branch is not viable.
